File: Reference/tests/Brontide.Reference.Interchange.Tests/Portable/PortableEstablishmentTests.cs
using System.Collections.Immutable;
using Brontide.Reference.Experimental.Binding.Portable;

namespace Brontide.Reference.Interchange.Tests.Portable;

/// <summary>Neutral vectors PB-01 through PB-17: contract establishment and the Shape floor.</summary>
public sealed class PortableEstablishmentTests
{
    private static readonly PortableShapeCatalog Catalog =
        PortableShapeCatalog.FromContract(CoolingPortableFixture.Contract);

    private static PortableBindingPlan Negotiate(
        PortableContractDocument? required = null,
        PortableContractDocument? offered = null) =>
        PortableNegotiation.Negotiate(
            required ?? CoolingPortableFixture.Contract,
            offered ?? CoolingPortableFixture.Contract,
            PortableRealization.FixedDirectCall,
            "host",
            "provider",
            "test");

    private static CborMap EncodedContract() =>
        (CborMap)PortableContractCodec.Encode(CoolingPortableFixture.Contract);

    private static PortableProtocolCategory CategoryOf(Action action) =>
        Assert.Throws<PortableFaultException>(action)!.Category;

    // PB-01-EXACT-ESTABLISHMENT
    [Test]
    public void Exact_establishment_freezes_a_plan_without_a_provider_effect()
    {
        var handler = new CoolingPortableHandler(CoolingPortableFixture.CreateNativeRegistry());
        var endpoint = PortableTestHarness.CoolingEndpoint(PortableRealization.FixedDirectCall, handler);
        var accepted = endpoint.Establish(CoolingPortableFixture.Contract, "host");

        Assert.Multiple(() =>
        {
            Assert.That(endpoint.Plan, Is.Not.Null);
            Assert.That(endpoint.Plan!.Operations, Does.Contain(CoolingPortableFixture.SetEnabled));
            Assert.That(accepted.CompactIdentifiers, Is.Not.Empty);
            Assert.That(handler.ProviderEffectCount, Is.Zero);
            Assert.That(endpoint.State, Is.EqualTo(PortableLifecycleState.Established));
        });
    }

    // PB-02-CONTRACT-VERSION-SKEW
    [Test]
    public void A_contract_version_the_endpoint_does_not_recognize_fails_closed()
    {
        var skewed = EncodedContract().With("contractVersion", new CborInteger(2));
        Assert.That(
            CategoryOf(() => PortableContractCodec.Decode(skewed)),
            Is.EqualTo(PortableProtocolCategory.UnsupportedVersion));
    }

    // PB-03-REQUIRED-REQUIREMENT-UNMET
    [Test]
    public void A_required_requirement_without_a_matching_provision_fails_closed()
    {
        Assert.That(
            CategoryOf(() => Negotiate(offered: CoolingPortableFixture.WithoutProfileProvision())),
            Is.EqualTo(PortableProtocolCategory.UnsupportedContract));
    }

    // PB-04-OPPOSED-REQUIREMENT-OFFERED
    [Test]
    public void An_opposed_requirement_that_is_offered_is_refused_rather_than_ignored()
    {
        Assert.That(
            CategoryOf(() => Negotiate(offered: CoolingPortableFixture.WithStreamingProvision())),
            Is.EqualTo(PortableProtocolCategory.UnsupportedContract));
    }

    // PB-05-UNKNOWN-CONTRACT-FIELD
    [Test]
    public void An_undeclared_contract_field_is_rejected_before_negotiation()
    {
        var extended = EncodedContract().With("hint", new CborText("ignore me"));
        Assert.That(
            CategoryOf(() => PortableContractCodec.Decode(extended)),
            Is.EqualTo(PortableProtocolCategory.MalformedMessage));
    }

    // PB-06-UNKNOWN-ENUMERATION-VALUE
    [Test]
    public void A_value_outside_a_declared_enumeration_is_not_defaulted_to_a_permissive_member()
    {
        var contract = EncodedContract();
        var requirements = contract.Array("requirements");
        var first = ((CborMap)requirements.Items[0]).With("strength", new CborText("mandatory"));
        Assert.That(
            CategoryOf(() => PortableContractCodec.Decode(contract.With("requirements", requirements.Replace(0, first)))),
            Is.EqualTo(PortableProtocolCategory.MalformedMessage));
    }

    // PB-07-COMPACT-IDENTIFIER-BEFORE-NEGOTIATION
    [Test]
    public void A_compact_identifier_this_binding_never_assigned_resolves_to_nothing()
    {
        var endpoint = PortableTestHarness.CoolingEndpoint(PortableRealization.FixedDirectCall);
        endpoint.Establish(CoolingPortableFixture.Contract, "host");
        endpoint.SignalReady();

        Assert.That(
            CategoryOf(() => endpoint.Request(
                PortableChannelRequestId.New(),
                null,
                4242,
                CoolingPortableFixture.CommandV1,
                CoolingPortableFixture.Command("primary", enabled: true),
                [])),
            Is.EqualTo(PortableProtocolCategory.UnsupportedContract));
    }

    // PB-08-NAME-OUTSIDE-PORTABLE-PROFILE
    [Test]
    public void A_name_outside_the_portable_profile_is_rejected_rather_than_kept_as_an_opaque_string()
    {
        Assert.Multiple(() =>
        {
            // A Unicode-letter name is a valid Brontide canonical name but is not portable in 0.1.
            Assert.That(PortableName.TryParse("interchange.tests.kühlung", out _), Is.False);
            Assert.That(PortableName.TryParse("a:b:c", out _), Is.False);
            Assert.That(PortableName.TryParse("interchange..tests", out _), Is.False);
            Assert.That(PortableName.TryParse("interchange.tests.cooling-component", out _), Is.True);
            Assert.That(
                CategoryOf(() => PortableName.Parse("interchange.tests.kühlung")),
                Is.EqualTo(PortableProtocolCategory.MalformedMessage));
        });
    }

    // PB-09-REQUEST-BEFORE-READY
    [Test]
    public void A_request_before_the_readiness_signal_produces_no_provider_effect()
    {
        var handler = new CoolingPortableHandler(CoolingPortableFixture.CreateNativeRegistry());
        var endpoint = PortableTestHarness.CoolingEndpoint(PortableRealization.FixedDirectCall, handler);
        endpoint.Establish(CoolingPortableFixture.Contract, "host");

        Assert.Multiple(() =>
        {
            Assert.That(
                CategoryOf(() => endpoint.Request(
                    PortableChannelRequestId.New(),
                    CoolingPortableFixture.SetEnabled,
                    null,
                    CoolingPortableFixture.CommandV1,
                    CoolingPortableFixture.Command("primary", enabled: true),
                    [])),
                Is.EqualTo(PortableProtocolCategory.StateViolation));
            Assert.That(handler.ProviderEffectCount, Is.Zero);
        });
    }

    // PB-10-INLINE-NESTED-AND-REPEATED-VALUES
    [Test]
    public async Task Nested_and_repeated_inline_values_preserve_sequence_order_exactly()
    {
        await using var host = await PortableTestHarness.DirectHostAsync(
            CatalogPortableFixture.Contract,
            new CatalogPortableHandler());

        var upsert = await host.InvokeAsync(
            CatalogPortableFixture.Upsert,
            CatalogPortableFixture.UpsertCommand,
            CatalogPortableFixture.UpsertCommandValue(
                CatalogPortableFixture.ItemValue("a", "Alpha", "one", "two"),
                CatalogPortableFixture.ItemValue("b", "Beta", "two", "one")),
            PortableTestHarness.Permitted(),
            [CatalogPortableFixture.Handle()]);
        Assert.That(upsert.ResultClass, Is.EqualTo(PortableResultClass.OutcomeSucceeded));

        var find = await host.InvokeAsync(
            CatalogPortableFixture.Find,
            CatalogPortableFixture.FindCommand,
            CatalogPortableFixture.FindCommandValue("b", "a"),
            PortableTestHarness.Permitted(),
            [CatalogPortableFixture.Handle()]);

        var items = (PortableSequenceValue)((PortableRecordValue)find.Value!).Fields["items"];
        var ids = items.Items
            .Cast<PortableRecordValue>()
            .Select(item => ((PortableTextValue)item.Fields["id"]).Value)
            .ToImmutableArray();
        Assert.That(ids, Is.EqualTo(new[] { "b", "a" }).AsCollection, "Sequence order is semantic and is preserved.");
    }

    // PB-11-ADDITIVE-PROJECTION and PB-57-OBSERVATION-RECORDS-MAPPING-OBLIGATIONS
    [Test]
    public async Task An_additive_payload_projects_and_the_projection_is_a_recorded_obligation()
    {
        await using var host = await PortableTestHarness.DirectHostAsync();
        var decision = host.PrepareRequest(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV2,
            PortableTestHarness.Permitted());

        Assert.Multiple(() =>
        {
            Assert.That(decision.FrameDecision, Is.EqualTo(PortableFrameDecision.Emit));
            Assert.That(decision.ResultClass, Is.EqualTo(PortableResultClass.Request));
            Assert.That(decision.MappingObligations, Has.One.Contains("projected:"));
        });

        var result = await host.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV2,
            CoolingPortableFixture.Command("primary", enabled: true, requestedBy: "operator-2"),
            PortableTestHarness.Permitted(),
            prepared: decision);

        Assert.Multiple(() =>
        {
            Assert.That(result.ResultClass, Is.EqualTo(PortableResultClass.OutcomeSucceeded));
            Assert.That(result.Observation.MappingObligations, Has.One.Contains("projected:"));
        });
    }

    // PB-12-NON-ADDITIVE-SKEW
    [Test]
    public async Task A_non_additive_Shape_version_is_refused_and_begins_no_provider_effect()
    {
        var handler = new CoolingPortableHandler(CoolingPortableFixture.CreateNativeRegistry());
        await using var host = await PortableTestHarness.DirectHostAsync(handler: handler);
        var result = await host.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV3,
            PortableRecordValue.Of(
                ("loop", new PortableTextValue("primary")),
                ("enabled", new PortableBooleanValue(true)),
                ("reason", new PortableTextValue("maintenance"))),
            PortableTestHarness.Permitted());

        Assert.Multiple(() =>
        {
            Assert.That(result.ResultClass, Is.EqualTo(PortableResultClass.ProtocolError));
            Assert.That(result.Category, Is.EqualTo(PortableProtocolCategory.InvalidPayload));
            Assert.That(result.Observation.ProviderEffectCount, Is.Zero);
            Assert.That(handler.ProviderEffectCount, Is.Zero);
        });
    }

    // PB-13-CLOSED-FRAGMENT-POLICY-VIOLATED
    [Test]
    public async Task A_closed_Shape_refuses_a_Fragment_the_Operation_does_not_declare()
    {
        await using var host = await PortableTestHarness.DirectHostAsync(handler: new NotingCoolingHandler());
        var result = await host.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true),
            PortableTestHarness.Permitted());

        Assert.Multiple(() =>
        {
            Assert.That(result.ResultClass, Is.EqualTo(PortableResultClass.ProtocolError));
            Assert.That(result.Category, Is.EqualTo(PortableProtocolCategory.InvalidPayload));
        });
    }

    // PB-14-NULL-IN-NON-UNIT-OPTIONAL-FIELD
    [Test]
    public void A_present_but_null_optional_field_is_refused_because_absence_is_omission()
    {
        var record = CborArray.Of(
            CborMap.Of(
            [
                ("loop", new CborText("primary")),
                ("enabled", new CborBoolean(true)),
                ("failureMode", CborNull.Instance)
            ]),
            CborMap.Empty);

        Assert.That(
            CategoryOf(() => PortableValueCodec.Decode(Catalog, CoolingPortableFixture.CommandV1, record)),
            Is.EqualTo(PortableProtocolCategory.InvalidPayload));
    }

    // PB-15-UNKNOWN-CHOICE-ALTERNATIVE
    [Test]
    public void An_alternative_outside_the_declared_set_is_an_invalid_payload_not_an_unknown_kind()
    {
        var choice = CborArray.Of(new CborText("elsewhere"), new CborText("value"));
        Assert.That(
            CategoryOf(() => PortableValueCodec.Decode(Catalog, CoolingPortableFixture.EncodingChoice, choice)),
            Is.EqualTo(PortableProtocolCategory.InvalidPayload),
            "The envelope kind was recognized, so the refusal belongs to the payload.");
    }

    // PB-16-EXCEPTION-SHAPED-PAYLOAD-DATA
    [Test]
    public void Foreign_runtime_identity_in_a_payload_position_is_refused_before_semantic_use()
    {
        var body = CborMap.Of(
        [
            ("operation", CborAccessTestDouble.Reference(CoolingPortableFixture.SetEnabled)),
            ("inputShape", CborAccessTestDouble.Reference(CoolingPortableFixture.CommandV1)),
            ("input", CborArray.Of(
                CborMap.Of([("loop", new CborText("primary")), ("stackTrace", new CborText("at Provider.Invoke"))]),
                CborMap.Empty)),
            ("resources", new CborArray([]))
        ]);

        Assert.That(
            CategoryOf(() => PortableRequestBody.Decode(body)),
            Is.EqualTo(PortableProtocolCategory.InvalidPayload));
    }

    // PB-17-EXCEPTION-SHAPED-CONTROL-DATA
    [Test]
    public void Foreign_runtime_identity_in_a_control_position_is_malformed()
    {
        var envelope = CborMap.Of(
        [
            ("contractVersion", new CborInteger(1)),
            ("kind", new CborText("ready")),
            ("channelId", new CborText("c1")),
            ("exception", new CborText("System.InvalidOperationException")),
            ("body", CborMap.Empty)
        ]);

        Assert.That(
            CategoryOf(() => PortableEnvelopeCodec.Decode(PortableCbor.Encode(envelope), PortableLimits.Declared)),
            Is.EqualTo(PortableProtocolCategory.MalformedMessage));
    }

    /// <summary>A provider whose result attaches a Fragment the negotiated Operation does not declare.</summary>
    private sealed class NotingCoolingHandler : IPortableOperationHandler
    {
        public PortableOperationEffect Invoke(
            PortableOperationReference operation,
            PortableValue input,
            IReadOnlyList<PortableResource> resources) =>
            PortableOperationEffect.Success(
                PortableRecordValue.Of(
                        ("loop", new PortableTextValue("primary")),
                        ("coolingEnabled", new PortableBooleanValue(true)),
                        ("revision", new PortableIntegerValue(1)),
                        ("providerEffectCount", new PortableIntegerValue(1)))
                    .WithFragment(CoolingPortableFixture.Note, ("note", new PortableTextValue("undeclared"))),
                1);
    }
}

/// <summary>Builds the structured canonical reference the control positions carry.</summary>
internal static class CborAccessTestDouble
{
    public static CborItem Reference(PortableOperationReference reference) =>
        CborMap.Of([("name", new CborText(reference.Name.Value)), ("version", new CborInteger(reference.Version))]);

    public static CborItem Reference(PortableShapeReference reference) =>
        CborMap.Of([("name", new CborText(reference.Name.Value)), ("version", new CborInteger(reference.Version))]);
}

File: Minimal/tests/Brontide.Minimal.Interchange.Tests/Portable/PortableEstablishmentTests.fs
namespace Brontide.Minimal.Interchange.Tests.Portable

open NUnit.Framework
open Brontide.Minimal.Binding.Portable

/// Small edits over decoded control items, so an adversarial document can be built exactly rather
/// than hand-typed as bytes.
[<AutoOpen>]
module CborEdits =

    let withEntry key value item =
        match item with
        | CborMap entries -> CborMap((entries |> List.filter (fun (name, _) -> name <> key)) @ [ key, value ])
        | other -> other

    let withoutEntry key item =
        match item with
        | CborMap entries -> CborMap(entries |> List.filter (fun (name, _) -> name <> key))
        | other -> other

    let entry key item =
        match item with
        | CborMap entries -> entries |> List.pick (fun (name, value) -> if name = key then Some value else None)
        | _ -> failwith $"'{key}' is not a member of a non-map item."

    let replaceAt index value item =
        match item with
        | CborArray items -> CborArray(items |> List.mapi (fun position current -> if position = index then value else current))
        | other -> other

/// C1 and C5: negotiation, the Shape floor, projection, and forbidden content.
[<TestFixture>]
type PortableEstablishmentTests() =

    let catalog = catalogOf CoolingFixture.contract
    let limits = PortableLimits.declared

    let negotiate required offered =
        PortableNegotiation.negotiate required offered Realization.FixedDirectCall "host" "provider" "fixed"

    let establishedEndpoint () =
        let endpoint =
            PortableProviderEndpoint(CoolingFixture.contract, CoolingHandler(), Realization.FixedDirectCall)

        expectOk (endpoint.Establish(CoolingFixture.contract, "host")) |> ignore
        endpoint

    [<Test>]
    member _.``PB-01 an exactly equal contract establishes and freezes the plan``() =
        let handler = CoolingHandler()
        let plan = expectOk (negotiate CoolingFixture.contract CoolingFixture.contract)

        assertAll (fun () ->
            Assert.That(BindingPlan.contractVersion plan, Is.EqualTo 1)
            shouldEqual [ CoolingFixture.setEnabled ] (BindingPlan.operations plan)
            Assert.That(List.length (BindingPlan.satisfiedRequirements plan), Is.EqualTo 2)
            // Negotiation completes before any provider activation, so nothing has been invoked.
            Assert.That(handler.ProviderEffectCount, Is.EqualTo 0L))

    [<Test>]
    member _.``PB-02 a contract version the endpoint does not recognize fails closed``() =
        let document =
            ContractCodec.encode CoolingFixture.contract |> withEntry "contractVersion" (CborInteger 2L)

        expectCategory ProtocolCategory.UnsupportedVersion (ContractCodec.decode document) |> ignore

    [<Test>]
    member _.``PB-03 an unmet required requirement fails closed``() =
        expectCategory
            ProtocolCategory.UnsupportedContract
            (negotiate CoolingFixture.contract (CoolingFixture.withoutProfileProvision ()))
        |> ignore

    [<Test>]
    member _.``PB-04 an opposed requirement that is offered is refused rather than ignored``() =
        expectCategory
            ProtocolCategory.UnsupportedContract
            (negotiate CoolingFixture.contract (CoolingFixture.withStreamingProvision ()))
        |> ignore

    [<Test>]
    member _.``PB-05 an undeclared contract field is refused before negotiation``() =
        let document =
            ContractCodec.encode CoolingFixture.contract |> withEntry "extension" (CborText "unknown")

        expectCategory ProtocolCategory.MalformedMessage (ContractCodec.decode document) |> ignore

    [<Test>]
    member _.``PB-06 an unknown enumeration value is refused rather than defaulted``() =
        let document = ContractCodec.encode CoolingFixture.contract
        let requirements = entry "requirements" document

        let mutated =
            match requirements with
            | CborArray(head :: tail) -> CborArray(withEntry "strength" (CborText "mandatory") head :: tail)
            | other -> other

        expectCategory
            ProtocolCategory.MalformedMessage
            (ContractCodec.decode (withEntry "requirements" mutated document))
        |> ignore

    [<Test>]
    member _.``PB-07 a compact identifier this binding never assigned resolves to no identity``() =
        let endpoint = establishedEndpoint ()

        expectCategory ProtocolCategory.UnsupportedContract (endpoint.ResolveOperation(OperationDesignation.Compact 4242))
        |> ignore

    [<Test>]
    member _.``PB-08 a name outside the portable profile is refused rather than accepted opaquely``() =
        assertAll (fun () ->
            // A Unicode-letter name is a valid Brontide canonical name but is not portable in 0.1.
            expectCategory ProtocolCategory.MalformedMessage (PortableName.tryCreate "interchange.tests.coöling")
            |> ignore

            expectCategory ProtocolCategory.MalformedMessage (PortableName.tryCreate "a:b:c") |> ignore
            expectCategory ProtocolCategory.MalformedMessage (PortableName.tryCreate "trailing.") |> ignore
            expectCategory ProtocolCategory.MalformedMessage (PortableCanonical.tryCreate "interchange.tests.ok" 0)
            |> ignore)

    [<Test>]
    member _.``PB-09 a request before the readiness signal is refused``() =
        let endpoint = establishedEndpoint ()

        let refused =
            endpoint.Request(
                ChannelRequestId.next (),
                OperationDesignation.Canonical CoolingFixture.setEnabled,
                CoolingFixture.commandV1,
                CoolingFixture.authorizedCommand "primary" true,
                []
            )

        expectCategory ProtocolCategory.StateViolation refused |> ignore

    [<Test>]
    member _.``PB-10 nested and repeated inline values conform and preserve order``() =
        let handler = CatalogHandler()
        let host = directHost CatalogFixture.contract handler

        let items =
            [ CatalogFixture.itemValue "a" "Alpha" [ "cold"; "cold"; "spare" ]
              CatalogFixture.itemValue "b" "Beta" [ "warm" ] ]

        let stored =
            invoke host CatalogFixture.upsert CatalogFixture.upsertCommand (CatalogFixture.upsertCommandValue items)

        Assert.That(stored.ResultClass, Is.EqualTo ResultClass.OutcomeSucceeded)

        let found =
            invoke host CatalogFixture.find CatalogFixture.findCommand (CatalogFixture.findCommandValue [ "b"; "a" ])

        Assert.That(found.ResultClass, Is.EqualTo ResultClass.OutcomeSucceeded)

        match found.Value |> Option.bind (PortableRecord.tryField "items") with
        | Some(PortableSequence [ first; second ]) ->
            // Sequence order is semantic: the answer follows the order the request asked in.
            Assert.That(PortableRecord.tryField "id" first, Is.EqualTo(Some(PortableText "b")))
            Assert.That(PortableRecord.tryField "id" second, Is.EqualTo(Some(PortableText "a")))

            match PortableRecord.tryField "tags" first with
            | Some(PortableSequence tags) -> Assert.That(List.length tags, Is.EqualTo 1)
            | _ -> Assert.Fail "The found item carries its declared tag sequence."
        | _ -> Assert.Fail "The find result carries the requested items in order."

    [<Test>]
    member _.``PB-11 an additive version difference is projected rather than refused``() =
        let host = directCoolingHost ()

        let command =
            CoolingFixture.command "primary" true None (Some "operator") (Some "supervisor")

        let result = invoke host CoolingFixture.setEnabled CoolingFixture.commandV2 command

        assertAll (fun () ->
            Assert.That(result.FrameDecision, Is.EqualTo FrameDecision.Accept)
            Assert.That(result.ResultClass, Is.EqualTo ResultClass.OutcomeSucceeded)
            Assert.That(result.Observation.ProviderEffectCount, Is.EqualTo 1L))

    [<Test>]
    member _.``PB-12 a non-additive version difference refuses projection``() =
        let host = directCoolingHost ()

        let command =
            PortableRecord.ofFields
                [ "loop", PortableText "primary"
                  "enabled", PortableBoolean true
                  "reason", PortableText "maintenance" ]
            |> PortableRecord.withFragment CoolingFixture.hostContext [ "requesterLabel", PortableText "operator" ]

        let result = invoke host CoolingFixture.setEnabled CoolingFixture.commandV3 command

        assertAll (fun () ->
            Assert.That(result.FrameDecision, Is.EqualTo FrameDecision.Reject)
            Assert.That(result.Category, Is.EqualTo(Some ProtocolCategory.InvalidPayload))
            Assert.That(result.Observation.ProviderEffectCount, Is.EqualTo 0L))

    [<Test>]
    member _.``PB-13 a closed result Shape refuses an undeclared Fragment rather than projecting it``() =
        let attaching =
            { new IPortableOperationHandler with
                member _.Invoke(_, _, _) =
                    Ok(
                        EffectSucceeded(
                            PortableRecord.ofFields
                                [ "loop", PortableText "primary"
                                  "coolingEnabled", PortableBoolean true
                                  "revision", PortableInteger 1L
                                  "providerEffectCount", PortableInteger 1L ]
                            |> PortableRecord.withFragment CoolingFixture.note [ "note", PortableText "extra" ],
                            1L
                        )
                    ) }

        let host = directHost CoolingFixture.contract attaching

        let result =
            invoke host CoolingFixture.setEnabled CoolingFixture.commandV1 (CoolingFixture.authorizedCommand "primary" true)

        assertAll (fun () ->
            Assert.That(result.FrameDecision, Is.EqualTo FrameDecision.Reject)
            Assert.That(result.Category, Is.EqualTo(Some ProtocolCategory.InvalidPayload)))

    [<Test>]
    member _.``PB-14 a present-but-null optional field is refused``() =
        // Absence is expressed by omission, and null belongs only to unit.
        let encoded =
            CborArray
                [ CborMap
                      [ "enabled", CborBoolean true
                        "failureMode", CborNull
                        "loop", CborText "primary" ]
                  CborMap [] ]

        expectCategory
            ProtocolCategory.InvalidPayload
            (PortableValueCodec.decode catalog CoolingFixture.commandV1 [] encoded)
        |> ignore

    [<Test>]
    member _.``PB-15 an alternative outside the declared set is a payload decision``() =
        let encoded = CborArray [ CborText "quantity"; CborInteger 3L ]

        let fault =
            expectCategory
                ProtocolCategory.InvalidPayload
                (PortableValueCodec.decode catalog CoolingFixture.encodingChoice [] encoded)

        // The envelope kind was recognized, so this is invalid-payload rather than unsupported-kind.
        Assert.That(fault.LocalCode, Is.EqualTo "unknown-alternative")

    [<Test>]
    member _.``PB-16 exception-shaped data in a payload position is refused``() =
        let body =
            RequestBody.encode
                { Operation = OperationDesignation.Canonical CoolingFixture.setEnabled
                  InputShape = CoolingFixture.commandV1
                  Input = CborArray [ CborMap [ "stackTrace", CborText "at Provider.Invoke" ]; CborMap [] ]
                  Resources = [] }

        expectCategory ProtocolCategory.InvalidPayload (RequestBody.decode body) |> ignore

    [<Test>]
    member _.``PB-17 exception-shaped data in a control position is malformed``() =
        let envelope =
            EnvelopeCodec.toItem (Envelope.empty EnvelopeKind.Ready (ChannelId.next ()))
            |> withEntry "exception" (CborText "System.InvalidOperationException")

        expectCategory ProtocolCategory.MalformedMessage (EnvelopeCodec.ofItem envelope) |> ignore

    [<Test>]
    member _.``PB-57 a projection is recorded as a mapping obligation``() =
        let host = directCoolingHost ()

        let command =
            CoolingFixture.command "primary" true None (Some "operator") (Some "supervisor")

        let result = invoke host CoolingFixture.setEnabled CoolingFixture.commandV2 command

        shouldEqual
            [ $"projected:{PortableShapeRef.text CoolingFixture.commandV2}->{PortableShapeRef.text CoolingFixture.commandV1}" ]
            result.Observation.MappingObligations

        // The provider side records which structure the projection discarded, which is what
        // distinguishes a representation mapping from a semantic translation.
        let endpoint =
            PortableProviderEndpoint(CoolingFixture.contract, CoolingHandler(), Realization.FixedDirectCall)

        expectOk (endpoint.Establish(CoolingFixture.contract, "host")) |> ignore
        expectOk (endpoint.SignalReady())

        let outcome =
            expectOk (
                endpoint.Request(
                    ChannelRequestId.next (),
                    OperationDesignation.Canonical CoolingFixture.setEnabled,
                    CoolingFixture.commandV2,
                    command,
                    []
                )
            )

        Assert.That(outcome.MappingObligations, Contains.Item "field-projected:requestedBy")

    [<Test>]
    member _.``the limits declaration is refused when it is internally inconsistent``() =
        let inconsistent =
            { PortableLimits.declared with
                MaxResourceBytes = limits.MaxByteStringBytes + 1 }

        expectCategory ProtocolCategory.MalformedMessage (PortableLimits.validate inconsistent) |> ignore
